File: crates/dry_run_core/src/history/filesystem_layout.rs
use crate::history::SnapshotKey;
use chrono::{DateTime, NaiveDateTime, TimeZone, Utc};
use std::path::{Path, PathBuf};

pub const SNAPSHOT_EXTENSION: &str = "json.zst";

const TS_FORMAT: &str = "%Y%m%dT%H%M%SZ";

#[must_use]
pub fn snapshot_path(
    root: &Path,
    key: &SnapshotKey,
    timestamp: DateTime<Utc>,
    content_hash: &str,
) -> PathBuf {
    root.join(&key.project_id.0)
        .join(&key.database_id.0)
        .join(format!(
            "{}-{}.{}",
            timestamp.format(TS_FORMAT),
            content_hash,
            SNAPSHOT_EXTENSION,
        ))
}

#[must_use]
pub fn parse_snapshot_filename(name: &str) -> Option<(DateTime<Utc>, String)> {
    let stem = name.strip_suffix(&format!(".{SNAPSHOT_EXTENSION}"))?;
    let (ts_str, hash) = stem.split_once('-')?;
    let naive = NaiveDateTime::parse_from_str(ts_str, TS_FORMAT).ok()?;
    let ts = Utc.from_utc_datetime(&naive);

    Some((ts, hash.to_string()))
}

File: internal/history/filesystem_layout.go
package history

import (
	"fmt"
	"path/filepath"
	"strings"
	"time"
)

// layout <root>/<project>/<database>/<ts>-<hash>.json.zst for FIleSystemStore
const (
	bundleTimeLayout = "20060102T150405Z"
	bundleExtension  = ".json.zst"
)

func BundleDir(root string, key SnapshotKey) string {
	return filepath.Join(root, string(key.ProjectID), string(key.DatabaseID))
}

func BundleFilename(ts time.Time, contentHash string) string {
	return fmt.Sprintf("%s-%s%s", ts.UTC().Format(bundleTimeLayout), contentHash, bundleExtension)
}

func BundlePath(root string, key SnapshotKey, ts time.Time, contentHash string) string {
	return filepath.Join(BundleDir(root, key), BundleFilename(ts, contentHash))
}

// inverse of BundleFilename; returns (ts, content_hash, ok)
func ParseBundleFilename(name string) (time.Time, string, bool) {
	if !strings.HasSuffix(name, bundleExtension) {
		return time.Time{}, "", false
	}
	stem := strings.TrimSuffix(name, bundleExtension)
	i := strings.IndexByte(stem, '-')
	if i < 0 || i+1 >= len(stem) {
		return time.Time{}, "", false
	}
	ts, err := time.Parse(bundleTimeLayout, stem[:i])
	if err != nil {
		return time.Time{}, "", false
	}
	return ts, stem[i+1:], true
}
